sled-agent/src/services.rs

```diff
@@ -1937,6 +1910,121 @@ impl ServiceManager {
+                                domain,
+                                ..
+                            },
+                        underlay_address,
+                        ..
+                    },
+                ..
+            })
+            | ZoneArgs::Omicron(OmicronZoneConfigLocal {
+                zone:
+                    OmicronZoneConfig {
+                        zone_type:
+                            OmicronZoneType::InternalNtp {
+                                dns_servers,
+                                ntp_servers,
+                                domain,
+                                ..
+                            },
+                        underlay_address,
+                        ..
+                    },
+                ..
+            }) => {
+                let Some(info) = self.inner.sled_info.get() else {
+                    return Err(Error::SledAgentNotReady);
+                };
+
+                let static_addr = underlay_address.to_string();
+
+                let nw_setup_service = Self::zone_network_setup_install(
+                    info,
+                    &installed_zone,
+                    &static_addr.clone(),
+                )?;
+
+                let is_boundary = matches!(
+                    // It's safe to unwrap here as we already know it's one of InternalNTP or BoundaryNtp.
+                    request.omicron_type().unwrap(),
+                    OmicronZoneType::BoundaryNtp { .. }
+                );
+
+                let rack_net =
+                    Ipv6Subnet::<RACK_PREFIX>::new(info.underlay_address)
+                        .net()
+                        .to_string();
+
+                let dns_install_service =
+                    Self::dns_install(info, Some(dns_servers.to_vec()), domain)
+                        .await?;
+
+                let mut ntp_config = PropertyGroupBuilder::new("config")
+                    .add_property("allow", "astring", &rack_net)
+                    .add_property(
+                        "boundary",
+                        "boolean",
+                        &is_boundary.to_string(),
+                    );
+
+                for s in ntp_servers {
+                    ntp_config = ntp_config.add_property(
+                        "server",
+                        "astring",
+                        &s.to_string(),
+                    );
+                }
+
+                let dns_client_service;
+                if dns_servers.is_empty() {
+                    dns_client_service =
+                        ServiceBuilder::new("network/dns/client").add_instance(
+                            ServiceInstanceBuilder::new("default").disable(),
+                        );
+                } else {
+                    dns_client_service = ServiceBuilder::new(
+                        "network/dns/client",
+                    )
+                    .add_instance(ServiceInstanceBuilder::new("default"));
```

Comment:
I think you should move this up and make it more generic. Any zone that configures DNS via `dns/install` should also enable `dns/client` (cockroach zones aren't currently doing that), and any zone that does not configure DNS should disable `dns/client`.

This is similar in concept to disabling ssh, which we should do in all zones apart from the switch zone.
I suggest declaring a disabled `dns_client_service` at the same place as `disabled_ssh_service`
```rust
        let dns_client_service = ServiceBuilder::new("network/dns/client")
            .add_instance(ServiceInstanceBuilder::new("default").disable());
```
and have all zones use `dns_client_service` as part of their profile, but call `dns_client_service.enable()` 
 (new method) for any zone that also uses a `dns_install_service`.

Hopefully that makes sense.
The `network/dns/client` service itself doesn't actually do anything apart from serve as a dependency for the name services milestone, and so it's useful as a dependency for things like ntp - see comment elsewhere.